### DISCUSSION TITLE
Testing Node-RED dashboard on Moddable Two

### DISCUSSION THREAD

**Discussion by rei-vilo** [phoddie/node-red-mcu#55] [Q&A]
After #40, I am testing a Node-RED dashboard on the Moddable Two board. I am facing a `XS abort: stack overflow` error.

<img width="419" alt="Capture 2022-12-14 à 13 47 26" src="https://user-images.githubusercontent.com/951776/207599148-5ba73486-65bb-4622-b778-08a954c0f138.png">

<details>

```
[
    {
        "id": "767a72570baf9755",
        "type": "ui_button",
        "z": "7c1235b869b51ef2",
        "name": "",
        "group": "bf78367624b9346f",
        "order": 0,
        "width": 0,
        "height": 0,
        "passthru": false,
        "label": "button",
        "tooltip": "",
        "color": "",
        "bgcolor": "",
        "className": "",
        "icon": "",
        "payload": "",
        "payloadType": "date",
        "topic": "topic",
        "topicType": "msg",
        "_mcu": {
            "mcu": true
        },
        "x": 210,
        "y": 100,
        "wires": [
            [
                "5d8a8590eefa70c9"
            ]
        ]
    },
    {
        "id": "5d8a8590eefa70c9",
        "type": "ui_text",
        "z": "7c1235b869b51ef2",
        "group": "bf78367624b9346f",
        "order": 1,
        "width": 0,
        "height": 0,
        "name": "",
        "label": "text",
        "format": "{{msg.payload}}",
        "layout": "row-spread",
        "className": "",
        "_mcu": {
            "mcu": true
        },
        "x": 490,
        "y": 160,
        "wires": []
    },
    {
        "id": "bf78367624b9346f",
        "type": "ui_group",
        "name": "Default",
        "tab": "4b8489d389053d4b",
        "order": 1,
        "disp": true,
        "width": "6",
        "collapse": false,
        "className": "",
        "_mcu": {
            "mcu": false
        }
    },
    {
        "id": "4b8489d389053d4b",
        "type": "ui_tab",
        "name": "Home",
        "icon": "dashboard",
        "disabled": false,
        "hidden": false,
        "_mcu": {
            "mcu": false
        }
    }
]
```

</details>

Build, link and upload go fine. 

However, at run time, **xsbug** reports `XS abort: stack overflow`

<img width="872" alt="207599326-f6d0d55c-2fd6-4681-9fc5-ef9f42da0873" src="https://user-images.githubusercontent.com/951776/207599771-fa978850-95f3-4f88-841c-8d696a80de3f.png">

What is going wrong?

Environment:

+ macOS 13.1
+ Node-RED 3.0.2 
+ Node-RED MCU plugin 1.0.0-rc.1
+ Fresh downloads of Moddable, ESP32 and Pico SDKs
+ Target esp32/moddable_two

**Comment by rei-vilo** [comment]
Does the selection of the board set the parameters of the display?

If no, what are the correct parameters for the Moddable Two display?

Thank you!

<img width="482" alt="Capture 2022-12-14 à 14 04 12" src="https://user-images.githubusercontent.com/951776/207602802-6ccf42aa-5c18-4fa1-973b-d9c595b5c21a.png"> <img width="481" alt="Capture 2022-12-14 à 14 03 56" src="https://user-images.githubusercontent.com/951776/207602815-f49d67df-41f8-4156-8f4e-f012d63c3b9b.png">

**Comment by rei-vilo** [comment]
Same issue when targeting the simulator for Moddable Two.

<img width="784" alt="Capture 2022-12-14 à 14 10 07" src="https://user-images.githubusercontent.com/951776/207604065-43c68b16-6f71-4746-ae27-de7066703460.png">

**Comment by rei-vilo** [comment]
Is there an official example featuring the Node-RED dashboard I could start with?

**Reply by ralphwetzel** [reply]
There are three "official" tests in the `node-red-mcu` repository:

- [switch](https://github.com/phoddie/node-red-mcu/tree/main/switch)
- [uitest](https://github.com/phoddie/node-red-mcu/tree/main/uitest)
- [weather](https://github.com/phoddie/node-red-mcu/tree/main/weather)

All build & run @ `Moddable Two` with the default settings of `UI Support` & the following `Creation Params`:

```json
{
    "static": 0,
    "chunk": {
        "initial": 65536,
        "incremental": 0
    },
    "heap": {
        "initial": 4096,
        "incremental": 0
    },
    "keys": {
        "available": 128
    },
    "stack": 600
}
```

**Reply by rei-vilo** [reply]
Thank you for the correct parameters. 

Examples `switch` and `uitkst` work fine.

For the `weather` example, as per the [documentation](https://github.com/phoddie/node-red-mcu#openweathermap),

> The API key is not exported by the Node-RED editor. Currently it must be entered manually in the source code of the weather node.

Adding `node.credentials.apikey = "API-KEY";` to /.node-red/node_modules/node-red-node-openweathermap:50 doesn't to work.

What is the correct procedure?

To be continued...

**Reply by phoddie** [reply]
To use the weather node, you need to modify this line in the implementation:

https://github.com/phoddie/node-red-mcu/blob/89ab68498571a76a0e74bfd998cfccca3c0456fd/nodes/openweathermap/weather.js#L30

You put your API  key in the string. You register for an OpenWeather API key [here:](https://openweathermap.org/appid)  (link from the Node-RED OpenWeather node's documentation)

**Reply by rei-vilo** [reply]
Great, thank you!
